modules/src/clients/ics07_tendermint/client_def.rs

```diff
@@ -185,116 +189,263 @@ impl ClientDef for TendermintClient {
+        receipt_path: &Path,
     ) -> Result<(), Ics02Error> {
-        todo!()
+        verify_delay_passed(ctx, client_state, connection_end)?;
+
+        verify_non_membership(
+            client_state,
+            connection_end.counterparty().prefix(),
+            proofs.object_proof(),
+            root,
+            receipt_path.to_string(),
+        )
     }
 
     fn verify_upgrade_and_update_state(
         &self,
         _client_state: &Self::ClientState,
         _consensus_state: &Self::ConsensusState,
-        _proof_upgrade_client: MerkleProof,
-        _proof_upgrade_consensus_state: MerkleProof,
+        _proof_upgrade_client: RawMerkleProof,
+        _proof_upgrade_consensus_state: RawMerkleProof,
     ) -> Result<(Self::ClientState, Self::ConsensusState), Ics02Error> {
         todo!()
     }
 }
 
+fn verify_membership(
+    client_state: &ClientState,
+    prefix: &CommitmentPrefix,
+    proof: &CommitmentProofBytes,
+    root: &CommitmentRoot,
+    path: String,
+    value: Vec<u8>,
+) -> Result<(), Ics02Error> {
+    let merkle_path = apply_prefix(prefix, vec![path]).map_err(Error::ics23_error)?;
+    let merkle_proof: MerkleProof = RawMerkleProof::try_from(proof.clone())
+        .map_err(Ics02Error::invalid_consensus_state_proof)?
+        .into();
+
+    merkle_proof
+        .verify_membership(
+            &client_state.proof_specs,
+            root.clone().into(),
+            merkle_path,
+            value,
+            0,
+        )
+        .map_err(|e| Ics02Error::tendermint(Error::ics23_error(e)))
+}
+
+fn verify_non_membership(
+    client_state: &ClientState,
+    prefix: &CommitmentPrefix,
+    proof: &CommitmentProofBytes,
+    root: &CommitmentRoot,
+    path: String,
+) -> Result<(), Ics02Error> {
+    let merkle_path = apply_prefix(prefix, vec![path]).map_err(Error::ics23_error)?;
+    let merkle_proof: MerkleProof = RawMerkleProof::try_from(proof.clone())
+        .map_err(Ics02Error::invalid_consensus_state_proof)?
+        .into();
+
+    merkle_proof
+        .verify_non_membership(&client_state.proof_specs, root.clone().into(), merkle_path)
+        .map_err(|e| Ics02Error::tendermint(Error::ics23_error(e)))
+}
+
+fn verify_delay_passed(
+    ctx: &dyn ChannelReader,
+    client_state: &ClientState,
+    connection_end: &ConnectionEnd,
+) -> Result<(), Ics02Error> {
+    let current_timestamp = ctx.host_timestamp();
+    let current_height = ctx.host_height();
+
+    let client_id = connection_end.client_id();
+    let processed_time = ctx
+        .processed_time(client_id)
+        .map_err(|_| Error::processed_time_not_found(client_id.clone()))?;
+    let processed_height = ctx
+        .processed_height(client_id)
+        .map_err(|_| Error::processed_height_not_found(client_id.clone()))?;
+
+    let delay_period_time = connection_end.delay_period();
+    // TODO get delay_period_height from delay_period_time
+    //let delay_period_height = get_block_delay(delay_period_time);
```

Comment:
ibc-go gets the parameter `MaxExpectedTimePerBlock` and calculates `delay_period_height` with `delay_period_time`.
I don't know how to calculate it in ibc-rs.

Reply:
I added `ChannelReader::max_expected_time_per_block()` to resolve this. IMHO, this should ideally be part of a `ParamReader` trait but let's leave it in `ChannelReader` for now. Furthermore, `ConnectionReader` could be a super trait of `ClientReader`, etc. I'll document these ideas in another issue.